src/main/java/com/google/step/YOUR_PROJECT_NAME_HERE/servlets/SearchServlet.java

```diff
@@ -15,24 +15,100 @@
+  private static final String CSE_URL =
+      "https://www.googleapis.com/customsearch/v1?key=" + API_KEY + "&cx=";
+
+  private final HttpClient httpClient =
+      HttpClient.newBuilder().version(HttpClient.Version.HTTP_2).build();
+
+  private static String encodeValue(String value) {
+    try {
+      return URLEncoder.encode(value, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
-    // Set some data on the request for the template to access.
-    request.setAttribute("previousQuery", request.getParameter("q"));
-    request.setAttribute("results", Arrays.asList("Aaron", "Iris", "Kayla"));
-    // Forward the request to the template (which is a servlet itself).
-    request.getRequestDispatcher("WEB-INF/templates/search.jsp").forward(request, response);
+    String query = "&q=" + encodeValue(request.getParameter("q"));
+    List<String> allLinks = new ArrayList<>();
+
+    // TODO: after implementing scraping, consider changing try/catch calls to a for-loop
+    /** Send request to retrieve card content through w3School site link from Google CSE */
+    try {
+      String w3Link = getLink(CSE_URL + W3_CSE_ID + query);
+      allLinks.add(w3Link);
+      // TODO: Call scraping function to return JSON card content
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+
+    /** Send request to retrieve card content through StackOverflow site link from Google CSE */
+    try {
+      String stackLink = getLink(CSE_URL + STACK_CSE_ID + query);
+      allLinks.add(stackLink);
+      // TODO: Call stackoverflow API to return JSON card content
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+
+    /** Send request to retrieve card content through GeeksForGeeks site link from Google CSE */
+    try {
+      String geeksLink = getLink(CSE_URL + GEEKS_CSE_ID + query);
+      allLinks.add(geeksLink);
+      // TODO: Call scraping function to return JSON card content
+    } catch (Exception e) {
+      e.printStackTrace();
+    }
+
+    response.setContentType("text/html;");
+    response.getWriter().println(allLinks);
+  }
+
+  private String getLink(String CSE) throws Exception {
+    HttpRequest linkRequest =
+        HttpRequest.newBuilder()
+            .GET()
+            .uri(URI.create(CSE))
+            .setHeader("User-Agent", "Java 11 HttpClient Bot")
+            .build();
+
+    try {
+      HttpResponse<String> linkResponse =
+          httpClient.send(linkRequest, HttpResponse.BodyHandlers.ofString());
+      /** Parse JSON of CSE SRP to retrieve link from only the first search result */
+      JSONObject obj = new JSONObject(linkResponse.body());
+      String link = obj.getJSONArray("items").getJSONObject(0).getString("link");
+      return link;
```

Comment:
[optional] I personally think APIs are much more ergonomic if it returns null in the failure case instead of throwing an exception. A null-check is typically more concise than a try-catch.